### PR TITLE
Simplify managing of build-js.sh

### PR DESCRIPTION
**Pull request by dcodeIO**
This PR aims to make it easier to deal with `build-js.sh` in that it

* Reuses the [new](https://github.com/WebAssembly/binaryen/pull/2342) `BINARYEN_API` define to apply `EMSCRIPTEN_KEEPALIVE` in Emscripten builds so we don't have to manage the list of exported functions manually
* Obtains the list of source files automatically so all we have to do is to exclude specific files, which is rare

### CODE REVIEW

**Comment by MaxGraey** [comment]
Btw as I remember if switch from `--closure 1` to `--closure 2` [here](https://github.com/WebAssembly/binaryen/blob/8ef3c6dc563aec6a794001208e4b617ad20e3abe/build-js.sh#L59) we could get much nicer bundle size (~10% less then before)

File: build-js.sh [hunk 1]
```diff
@@ -73,876 +72,49 @@ OUT="$PWD/out"
 
 echo "generate embedded intrinsics module"
 
-python3 $BINARYEN_SCRIPTS/embedwast.py $BINARYEN_SRC/passes/wasm-intrinsics.wast $BINARYEN_SRC/passes/WasmIntrinsics.cpp
+python $BINARYEN_SCRIPTS/embedwast.py $BINARYEN_SRC/passes/wasm-intrinsics.wast $BINARYEN_SRC/passes/WasmIntrinsics.cpp
```
Comment (sbc100) [inline build-js.sh]: This change seems unrelated.

File: build-js.sh [hunk 2]
```diff
@@ -42,11 +43,9 @@ EMCC_ARGS="$EMCC_ARGS -s ALLOW_MEMORY_GROWTH=1"
 EMCC_ARGS="$EMCC_ARGS -s DEMANGLE_SUPPORT=1"
 EMCC_ARGS="$EMCC_ARGS -s NO_FILESYSTEM=0"
 EMCC_ARGS="$EMCC_ARGS -s WASM=0"
-EMCC_ARGS="$EMCC_ARGS -s ERROR_ON_UNDEFINED_SYMBOLS=1"
 EMCC_ARGS="$EMCC_ARGS -s BINARYEN_ASYNC_COMPILATION=0"
 EMCC_ARGS="$EMCC_ARGS -s ERROR_ON_UNDEFINED_SYMBOLS=1"
 EMCC_ARGS="$EMCC_ARGS -s DISABLE_EXCEPTION_CATCHING=0" # Exceptions are thrown and caught when optimizing endless loops
-OUT_FILE_SUFFIX=
```
Comment (sbc100) [inline build-js.sh]: Any idea what this used for?

File: build-js.sh [hunk 3]
```diff
@@ -42,11 +43,9 @@ EMCC_ARGS="$EMCC_ARGS -s ALLOW_MEMORY_GROWTH=1"
 EMCC_ARGS="$EMCC_ARGS -s DEMANGLE_SUPPORT=1"
 EMCC_ARGS="$EMCC_ARGS -s NO_FILESYSTEM=0"
 EMCC_ARGS="$EMCC_ARGS -s WASM=0"
-EMCC_ARGS="$EMCC_ARGS -s ERROR_ON_UNDEFINED_SYMBOLS=1"
 EMCC_ARGS="$EMCC_ARGS -s BINARYEN_ASYNC_COMPILATION=0"
 EMCC_ARGS="$EMCC_ARGS -s ERROR_ON_UNDEFINED_SYMBOLS=1"
```
Comment (sbc100) [inline build-js.sh]: This is the default so it can be removed.

File: build-js.sh [hunk 4]
```diff
@@ -73,876 +72,49 @@ OUT="$PWD/out"
...
+  "$BINARYEN_SRC/support/command-line.cpp"
+  "$BINARYEN_SRC/support/path.cpp"
+  "$BINARYEN_SRC/tools/"
+)
+
+# external source files to include
+INCLUDE=()
+
+mapfile -t SOURCES < <(ls $BINARYEN_SRC/**/*.cpp)
+for i in ${!SOURCES[@]}; do
```
Comment (sbc100) [inline build-js.sh]: I'm not huge find of bash script once they start to use these kinds of advanced features... but I think we can make an exception.

I wonder if we can't just use cmake here instead.. but that is a separate issue.

File: build-js.sh [hunk 5]
```diff
@@ -73,876 +72,49 @@ OUT="$PWD/out"
...
+  "$BINARYEN_SRC/support/archive.cpp"
+  "$BINARYEN_SRC/support/command-line.cpp"
+  "$BINARYEN_SRC/support/path.cpp"
+  "$BINARYEN_SRC/tools/"
+)
+
+# external source files to include
+INCLUDE=()
+
+mapfile -t SOURCES < <(ls $BINARYEN_SRC/**/*.cpp)
```
Comment (sbc100) [inline build-js.sh]: How about using `find -name "*.cpp"`?

File: build-js.sh (hunk 2 above)
Comment (dcodeIO) [reply build-js.sh]: Iirc it used to be different for debug builds, making it `binaryen-something.js`, but isn't used anymore apparently so I removed it.

File: build-js.sh (hunk 1 above)
Comment (dcodeIO) [reply build-js.sh]: Ah, right, needed to change this so it runs within my Debian VM since there's no python3 apparently. Do you want me to change it back?

File: build-js.sh (hunk 4 above)
Comment (dcodeIO) [reply build-js.sh]: On the binaryen.js bot I've been using [this node.js script](https://github.com/AssemblyScript/binaryen.js/blob/master/scripts/build.js) to do about the same, but recently switched back to utilize build-js.sh there, in turn leading to this PR. If there's interest, I can try to port the JS script over to the binaryen repo instead, and maybe add a `package.json` so one can run `npm run build`?

File: build-js.sh (hunk 5 above)
Comment (dcodeIO) [reply build-js.sh]: Are you proposing the following?

```
mapfile -t SOURCES < <(find $BINARYEN_SRC -name "*.cpp")
```

File: build-js.sh (hunk 4 above)
Comment (sbc100) [reply build-js.sh]: I'd rather just make the cmake build work.  I don't see why emscripten should be able to build binaryen in the normal way.. using cmake + ninja.

File: build-js.sh (hunk 1 above)
Comment (sbc100) [reply build-js.sh]: yes please.  If we are going to change python versions we should do it deliberately in its own PR.

File: build-js.sh (hunk 5 above)
Comment (sbc100) [reply build-js.sh]: Yeah, I'm just more familiar with that than the `**` trick in ls.. not a huge deal.. as I say hopefully we can remove this completely and switch the cmake.

File: build-js.sh (hunk 4 above)
Comment (dcodeIO) [reply build-js.sh]: I'm not familiar with cmake and ninja unfortunately. So far I've appreciated that I can run `bash build-js.sh` and it "just works" even on Windows or inside of a docker container that might not have the full set of build tools.

File: build-js.sh [hunk 6]
```diff
@@ -77,872 +75,46 @@ python3 $BINARYEN_SCRIPTS/embedwast.py $BINARYEN_SRC/passes/wasm-intrinsics.wast
...
+  "$BINARYEN_SRC/support/command-line.cpp"
+  "$BINARYEN_SRC/support/path.cpp"
+  "$BINARYEN_SRC/tools/"
+)
+
+# external source files to include
+INCLUDE=()
+
+readarray -d '' -t SOURCES < <(find $BINARYEN_SRC -name "*.cpp" -print0 | sort -z)
+EXCLUDE+=("$BINARYEN_SRC/binaryen-c.cpp")
```
Comment (sbc100) [inline build-js.sh]: Why this line?   Isn't this the main file we do want to include?

Comment (dcodeIO) [reply build-js.sh]: Yeah, but the file is excluded when compiling `shared.o`, and only included in the final step.
